src/extensionsIntegrated/CSSColorPreview/main.js

```diff
@@ -289,11 +257,128 @@ define(function (require, exports, module) {
+                        color: colorMatch[0],
+                        index: colorIndex
+                    });
+                }
+            });
+        }
+
+        // Return up to 4 colors
+        return validColors.slice(0, 4).map(item => item.color);
+    }
+
+    /**
+     * Responsible to get all the colors and their respective line numbers.
+     *
+     * @param {Editor} editor
+     * @return {Array.<Object>} an array of objects with all the line nos and,
+     *  the colors to be added on those lines
+     */
+    function _getAllColorsAndLineNums(editor) {
+        const nLen = editor.lineCount();
+        const aColors = [];
+
+        // Match colors and push into an array
+        for (let i = 0; i < nLen; i++) {
+            const colors = detectValidColorsInLine(editor, i);
+
+            // If valid colors found, add to the results
+            if (colors.length > 0) {
+                aColors.push({
+                    lineNumber: i,
+                    colorValues: colors
+                });
+            }
+        }
+
+        return aColors;
+    }
+
+
+    /**
+     * Responsible to update the color marks only on the modified lines
+     *
+     * @param {Editor} editor the editor instance
+     * @param {Number} fromLineNumber modification start from line number
+     * @param {Number} toLineNumber modification upto line number
+     * @return {Array.<Object>} an array of objects with all the line nos and,
+     *  the colors to be added on those lines
+     */
+    function updateColorMarks(editor, fromLineNumber, toLineNumber) {
+        const aColors = [];
+
+        // Match colors and push into an array for modified lines
+        for (let i = fromLineNumber; i <= toLineNumber; i++) {
+            const colors = detectValidColorsInLine(editor, i);
+
+            // If no valid colors, clear the gutter marker
+            if (colors.length === 0) {
+                editor.setGutterMarker(i, GUTTER_NAME, "");
+            } else {
+                aColors.push({
+                    lineNumber: i,
+                    colorValues: colors
+                });
+            }
+        }
+
+        return aColors;
+    }
+
+
     /**
      * Function that gets triggered when any change occurs on the editor
+     *
+     * @param {Editor} instance the codemirror instance
+     * @param {Object} changeObj an object that has properties regarding the line changed and type of change
      */
-    function onChanged() {
-        showColorMarks();
+    function onChanged(instance, changeObj) {
+
+        const editor = EditorManager.getActiveEditor();
+
+        // for insertion and deletion, update the changed lines
+        if(changeObj.origin === '+input' || changeObj.origin === '+delete') {
+            // make sure that the required properties exist and in the form they are expected to be
+            if(changeObj.from.line && changeObj.to.line && changeObj.from.line <= changeObj.to.line) {
```

Comment:
Check cases where there are colors at the end of the file and delete like 10 lines and see if there are no issues in console. Since the contents are deleted, will it cause any issues as the lines in the change object doesn't exist in the file anymore, what would be the behavior in that case.

Reply:
Fixed! Solution :- just add a check to make sure that lineText is not null. if it is null, then we can return an empty array.